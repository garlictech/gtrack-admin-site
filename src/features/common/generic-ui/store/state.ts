export interface State {
  progressSpinnerOn: boolean;
  progressSpinnerText?: string;
}

export const featureName = 'features.generic-ui';
