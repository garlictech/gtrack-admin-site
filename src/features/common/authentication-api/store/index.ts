import * as Actions from './actions';

export type Action = Actions.AllActions;
export { Effects } from './effects';

export { AuthenticationState, Auth, domain } from './state';
export { reducer as Reducer } from './reducer';
export { Actions };
