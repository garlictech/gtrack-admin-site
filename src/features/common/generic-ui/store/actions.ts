// tslint:disable:no-property-initializers max-classes-per-file
import { Action } from '@ngrx/store';
import { Toast } from '../interfaces';

export enum ActionTypes {
  ShowProgressSpinner = '[UI] Show progress spinner',
  HideProgressSpinner = '[UI] Hide progress spinner',
  DisplayToast = '[UI] display toast'
}

/**
 * The action displays a progress spinner hiding the full screen.
 *
 * @export
 */
export class ShowProgressSpinner implements Action {
  readonly type = ActionTypes.ShowProgressSpinner;
  /**
   * Creates an instance of ShowProgressSpinner.
   * @param [textLabel] A translateable, optional label that appears below the spinner. Use to add some
   * extra info about what is going on.
   */
  constructor(public textLabel?: string) {
    // EMPTY
  }
}

/**
 * Hide the progress spinner displeyed with {@link ShowProgressSpinner}.
 *
 * @export
 */
export class HideProgressSpinner implements Action {
  readonly type = ActionTypes.HideProgressSpinner;
}

/**
 * Display a toast message.
 *
 * @export
 */
export class DisplayToast implements Action {
  readonly type = ActionTypes.DisplayToast;
  /**
   * Creates an instance of DisplayToast.
   * @param notification The toast descriptor.
   */
  constructor(public notification: Toast) {
    // EMPTY
  }
}

export type AllActions = ShowProgressSpinner | HideProgressSpinner | DisplayToast;
