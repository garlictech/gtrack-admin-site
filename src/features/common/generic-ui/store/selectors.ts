import { createFeatureSelector, createSelector } from '@ngrx/store';

import { featureName, State } from './state';

export const selectFeature = createFeatureSelector<State>(featureName);

export const selectProgressSpinnerOn = createSelector(
  selectFeature,
  (state: State) => state.progressSpinnerOn
);
export const selectProgressSpinnerText = createSelector(
  selectFeature,
  (state: State) => state.progressSpinnerText
);
