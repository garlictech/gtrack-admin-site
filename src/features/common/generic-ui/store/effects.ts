// tslint:disable:no-property-initializers
import { Inject, Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { TranslateService } from '@ngx-translate/core';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';
import { EToastSeverity, GENERIC_UI_PLATFORM_SERVICE, GenericUiPlatformServiceData } from '../interfaces';
import * as fromActions from './actions';

@Injectable()
export class Effects {
  @Effect({ dispatch: false }) showProgressSpinner$: Observable<Action> = this.actions$.pipe(
    ofType<fromActions.ShowProgressSpinner>(fromActions.ActionTypes.ShowProgressSpinner),
    tap((action: fromActions.ShowProgressSpinner) => this._ps.displayLoader(action.textLabel))
  );

  @Effect({ dispatch: false }) hideProgressSpinner$: Observable<Action> = this.actions$.pipe(
    ofType(fromActions.ActionTypes.HideProgressSpinner),
    tap(() => this._ps.dismissLoader())
  );

  @Effect({ dispatch: false }) displayToast$: Observable<Action> = this.actions$.pipe(
    ofType<fromActions.DisplayToast>(fromActions.ActionTypes.DisplayToast),
    tap((action: fromActions.DisplayToast) =>
      this._ps.displayToast({
        severity: action.notification.severity || EToastSeverity.Success,
        summary: this._translate.instant(action.notification.summary),
        detail: action.notification.detail ? this._translate.instant(action.notification.detail) : undefined
      })
    )
  );
  constructor(
    private readonly actions$: Actions,
    @Inject(GENERIC_UI_PLATFORM_SERVICE) private readonly _ps: GenericUiPlatformServiceData,
    private readonly _translate: TranslateService
  ) {
    // EMPTY
  }
}
