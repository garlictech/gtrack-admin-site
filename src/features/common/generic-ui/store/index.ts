import * as Actions from './actions';

export type Action = Actions.AllActions;
export { Effects } from './effects';

export { State } from './state';
export * from './reducer';
export { Actions };
export * from './selectors';
