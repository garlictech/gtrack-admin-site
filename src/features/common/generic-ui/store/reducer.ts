import * as ApiActions from './actions';
import { State } from './state';

export const initialState: State = {
  progressSpinnerOn: false
};

export type Action = ApiActions.AllActions;

// tslint:disable-next-line:only-arrow-functions
export function reducer(state = initialState, action: Action): State {
  switch (action.type) {
    case ApiActions.ActionTypes.ShowProgressSpinner:
      return { ...state, progressSpinnerOn: true, progressSpinnerText: action.textLabel };

    case ApiActions.ActionTypes.HideProgressSpinner:
      return { ...state, progressSpinnerOn: false, progressSpinnerText: undefined };

    default:
      return state;
  }
}
