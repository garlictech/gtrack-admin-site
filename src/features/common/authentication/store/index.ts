// tslint:disable: only-arrow-functions
import { InjectionToken } from '@angular/core';
import { ActionReducer, ActionReducerMap } from '@ngrx/store';

import * as Actions from './actions';
import { reducer } from './reducer';
import * as Selectors from './selectors';

import { AuthenticationState } from './state';

export * from './state';
export { metaReducers } from './reducer';
export { Effects } from './effects';
export { Actions as AuthenticationActions };
export { Selectors as AuthenticationSelectors };

export function getReducers(): ActionReducer<AuthenticationState> {
  return reducer;
}

export const AUTHENTICATION_REDUCER_TOKEN = new InjectionToken<ActionReducerMap<AuthenticationState>>(
  'Registered Authentication Reducers'
);
