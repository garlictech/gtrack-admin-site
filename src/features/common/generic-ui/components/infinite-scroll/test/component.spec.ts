import { ChangeDetectorRef, Component, CUSTOM_ELEMENTS_SCHEMA, ElementRef, Injector } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { createSelector, Store, StoreModule } from '@ngrx/store';
import { of } from 'rxjs';
import { tap } from 'rxjs/operators';
import { InfiniteScrollComponent } from '../';
import { reducer } from '../../../store';

class MockElementRef {
  nativeElement = {};
}

describe('NativeInfiniteScrollComponent', () => {
  let component: TestInfiniteScrollComponent;
  let fixture: ComponentFixture<TestInfiniteScrollComponent>;
  const lastKeySelector = createSelector(state => 'foolastKey');
  const items$ = of(['foocollectionItem1', 'foocollectionItem2']);
  const itemFetcher = jest.fn();

  @Component({ selector: 'app-test-infinite-scroll', template: '' })
  class TestInfiniteScrollComponent extends InfiniteScrollComponent {
    _destroy = jest.fn();
    constructor(_injector: Injector) {
      super(_injector);
    }
  }

  let store;
  let elementRef: ElementRef;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [StoreModule.forRoot(reducer)],
      declarations: [TestInfiniteScrollComponent],
      providers: [{ provide: ChangeDetectorRef, useValue: {} }, { provide: ElementRef, useClass: MockElementRef }],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    }).compileComponents();

    store = TestBed.get(Store);
    elementRef = TestBed.get(ElementRef);
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestInfiniteScrollComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  afterEach(() => {
    component.ngOnDestroy();
    fixture.destroy();
  });

  it('it should be created', () => {
    expect(component).toBeTruthy();
  });

  it('ngOnDestroy should call _destroy', () => {
    component.ngOnDestroy();
    expect(component._destroy).toHaveBeenCalled();
  });

  it('initalize should initialize the component properly', done => {
    const sliceEmitterSpy = jest.spyOn(component.sliceEmitter$, 'next');
    component.initialize(lastKeySelector, items$, itemFetcher);
    expect(sliceEmitterSpy).toHaveBeenCalledWith({ firstIndex: 0, rows: 10 });
    expect(component.itemsShown).toMatchSnapshot('itemsShown');

    component.totalRecords$
      .pipe(
        tap(value => {
          expect(value).toMatchSnapshot('totalRecords$');
        })
      )
      .subscribe(() => done());
  });

  it('paginate must proceed to the next page', () => {
    const sliceEmitterSpy = jest.spyOn(component.sliceEmitter$, 'next');
    component.initialize(lastKeySelector, items$, itemFetcher);
    component.loadMore();
    expect(sliceEmitterSpy).toHaveBeenCalledWith({ firstIndex: 0, rows: 10 });
    expect(itemFetcher).toHaveBeenCalledWith('foolastKey');
  });

  it('totalRecords$ must return the total records if there are pages (lastKey is defined)', done => {
    component.initialize(lastKeySelector, items$, itemFetcher);

    component.totalRecords$.subscribe(recordNum => {
      // Because, items$ by default returns a 2-element array (see the fixtures), the total record number is 2 in this test.
      expect(recordNum).toMatchSnapshot();
      done();
    });
  });

  it('totalRecords$ must return the total records if there are NO pages (lastKey is undefined)', done => {
    const fLastkeySelector = state => undefined;
    component.initialize(fLastkeySelector, items$, itemFetcher);

    component.totalRecords$.subscribe(recordNum => {
      expect(recordNum).toMatchSnapshot();
      done();
    });
  });

  it('totalRecords$ must return 0 if there are NO items returned (undefined)', done => {
    const fLastkeySelector = state => undefined;
    const fitems$ = of(undefined);
    component.initialize(fLastkeySelector, fitems$, itemFetcher);

    component.totalRecords$.subscribe(recordNum => {
      expect(recordNum).toMatchSnapshot();
      done();
    });
  });

  it('it should emit the new items and slices when we paginate', () => {
    component.initialize(lastKeySelector, items$, itemFetcher);
    component.sliceEmitter$.next({ firstIndex: 0, rows: 1 });
    expect(component.itemsShown).toMatchSnapshot('itemsShown');
  });

  it('reset should emit sliceemitter with default properties', () => {
    const spy = spyOn(component.sliceEmitter$, 'next');
    component.reset();
    expect(spy).toHaveBeenCalledWith({ firstIndex: 0, rows: 10 });
  });

  it('shouldShowComponent$ should return true if there are more items', done => {
    const fLastkeySelector = state => 'lastkey';
    const fitems$ = of(undefined);
    component.initialize(fLastkeySelector, fitems$, itemFetcher);

    component.shouldShowComponent$.subscribe(shouldShow => {
      expect(shouldShow).toMatchSnapshot();
      done();
    });
  });
});
