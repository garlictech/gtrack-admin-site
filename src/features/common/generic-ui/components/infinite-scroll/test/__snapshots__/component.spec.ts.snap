// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NativeInfiniteScrollComponent initalize should initialize the component properly: itemsShown 1`] = `
Array [
  "foocollectionItem1",
  "foocollectionItem2",
]
`;

exports[`NativeInfiniteScrollComponent initalize should initialize the component properly: totalRecords$ 1`] = `2`;

exports[`NativeInfiniteScrollComponent it should emit the new items and slices when we paginate: itemsShown 1`] = `
Array [
  "foocollectionItem1",
  "foocollectionItem2",
]
`;

exports[`NativeInfiniteScrollComponent shouldShowComponent$ should return true if there are more items 1`] = `true`;

exports[`NativeInfiniteScrollComponent totalRecords$ must return 0 if there are NO items returned (undefined) 1`] = `0`;

exports[`NativeInfiniteScrollComponent totalRecords$ must return the total records if there are NO pages (lastKey is undefined) 1`] = `2`;

exports[`NativeInfiniteScrollComponent totalRecords$ must return the total records if there are pages (lastKey is defined) 1`] = `2`;
